schema/function_test.go

```diff
@@ -0,0 +1,517 @@
+	// Validate type
+	typeOutput, err := simpleFunc.Output(make([]schema.Type, 0))
+	assert.NoError(t, err)
+	assert.Equals[schema.Type](t, typeOutput, schema.NewIntSchema(nil, nil, nil))
+	// Validate that the function call works as intended
+	result, err := simpleFunc.Call([]any{"test"})
+	assert.NoError(t, err)
+	assert.Equals(t, result, 5)
+}
+
+func TestNewDynamicFunction_SameType(t *testing.T) {
+	simpleFunc, err := schema.NewDynamicCallableFunction(
+		"test",
+		[]schema.Type{schema.NewAnySchema()},
+		nil,
+		func(a any) (any, error) {
+			return a, nil
+		},
+		func(inputTypes []schema.Type) (schema.Type, error) {
+			if len(inputTypes) != 1 {
+				return nil, fmt.Errorf("expected 1 arg, got %d", len(inputTypes))
+			}
+			return inputTypes[0], nil
+		},
+	)
+	assert.NoError(t, err)
+
+	// Validate type
+	typeOutput, err := simpleFunc.Output([]schema.Type{schema.NewStringSchema(nil, nil, nil)})
+	assert.NoError(t, err)
+	assert.Equals[schema.Type](t, typeOutput, schema.NewStringSchema(nil, nil, nil))
+	// Validate that the function call works as intended
+	result, err := simpleFunc.Call([]any{"test"})
+	assert.NoError(t, err)
+	assert.Equals(t, result, "test")
+
+	// Validate type
+	typeOutput, err = simpleFunc.Output([]schema.Type{schema.NewIntSchema(nil, nil, nil)})
+	assert.NoError(t, err)
+	assert.Equals[schema.Type](t, typeOutput, schema.NewIntSchema(nil, nil, nil))
+	// Validate that the function call works as intended
+	result, err = simpleFunc.Call([]any{1})
+	assert.NoError(t, err)
+	assert.Equals(t, result, 1)
+
+	// Validate type
+	typeOutput, err = simpleFunc.Output([]schema.Type{schema.NewFloatSchema(nil, nil, nil)})
+	assert.NoError(t, err)
+	assert.Equals[schema.Type](t, typeOutput, schema.NewFloatSchema(nil, nil, nil))
+	// Validate that the function call works as intended
+	result, err = simpleFunc.Call([]any{5.0})
+	assert.NoError(t, err)
+	assert.Equals(t, result, 5.0)
+
+	// Validate type
+	typeOutput, err = simpleFunc.Output([]schema.Type{
+		schema.NewListSchema(schema.NewIntSchema(nil, nil, nil), nil, nil),
+	})
+	assert.NoError(t, err)
+	assert.Equals[schema.Type](t, typeOutput, schema.NewListSchema(schema.NewIntSchema(nil, nil, nil), nil, nil))
+	// Validate that the function call works as intended
+	result, err = simpleFunc.Call([]any{make([]int, 0)})
+	assert.NoError(t, err)
+	assert.Equals(t, result.([]int), make([]int, 0))
+}
+
+func TestNewDynamicFunction_SliceOfSameType(t *testing.T) {
+	simpleFunc, err := schema.NewDynamicCallableFunction(
+		"test",
+		[]schema.Type{schema.NewAnySchema()},
+		nil,
+		func(a any) (any, error) {
+			aVal := reflect.ValueOf(a)
+			result := reflect.MakeSlice(reflect.SliceOf(aVal.Type()), 2, 2)
+			result.Index(0).Set(aVal)
+			result.Index(1).Set(aVal)
+			return result.Interface(), nil
+		},
+		func(inputTypes []schema.Type) (schema.Type, error) {
+			if len(inputTypes) != 1 {
+				return nil, fmt.Errorf("expected 1 arg, got %d", len(inputTypes))
+			}
```

Comment:
I don't think so. What's there will just help in case someone changes the test incorrectly.

Reply:
In that case, shouldn't the result be a little "firmer"?  Perhaps instead of returning, it should fail an assertion or otherwise crash the test?  (Depending on the test itself to notice the error returned from here is somewhat weak.)

Reply:
Sure. Switched to assertion.